monitoring/opentelemetry/src/test/java/io/quarkus/ts/opentelemetry/OpenTelemetryMetricsGrpcTimeoutIT.java

```diff
@@ -0,0 +1,158 @@
+            .withProperty("quarkus.otel.metrics.enabled", "true")
+            .withProperty("quarkus.otel.exporter.otlp.metrics.endpoint", "http://localhost:" + METRICS_COLLECTOR_PORT)
+            .withProperty("quarkus.otel.exporter.otlp.metrics.protocol", "grpc")
+            .withProperty("quarkus.otel.metric.export.interval", EXPORT_INTERVAL_MS + "ms")
+            .withProperty("quarkus.otel.exporter.otlp.metrics.timeout", CONFIGURED_TIMEOUT_MS + "ms")
+            .withProperty("quarkus.otel.traces.enabled", "true")
+            .withProperty("quarkus.otel.logs.enabled", "false");
+
+    @Test
+    public void testGrpcTimeoutIsApplied(Vertx vertx) throws Exception {
+        resetCounters();
+
+        try (AutoCloseable collector = createFakeOtlpCollector(vertx)) {
+            LOG.info("Verifying normal metrics export works");
+            generateMetrics(2);
+
+            await().atMost(20, TimeUnit.SECONDS)
+                    .untilAsserted(() -> assertTrue(SUCCESSFUL_EXPORTS.get() > 0,
+                            "Should have at least one successful export"));
+
+            int baselineExports = SUCCESSFUL_EXPORTS.get();
+            LOG.infof("Baseline successful exports: %d", baselineExports);
+
+            LOG.info("Enabling request hanging to test timeout");
+            HANG_REQUESTS.set(true);
+
+            generateMetrics(1);
+
+            // Wait for the export to be attempted
+            await().atMost(10, TimeUnit.SECONDS)
+                    .untilAsserted(() -> assertTrue(HANGING_REQUESTS.get() > 0,
+                            "Should have at least one hanging request"));
+
+            LOG.infof("Hanging requests initiated: %d", HANGING_REQUESTS.get());
+
+            // With the fix the request should will timeout after 3s
+            LOG.info("Waiting for timeout to occur...");
+
+            long totalWaitTime = CONFIGURED_TIMEOUT_MS + (MAX_ATTEMPTS * RETRY_DELAY_MS) + 1000;
+            LOG.infof("Waiting %dms for timeout and retries to complete", totalWaitTime);
+            Thread.sleep(totalWaitTime);
+
+            LOG.info(" Disabling hanging to test recovery");
+            HANG_REQUESTS.set(false);
+
+            generateMetrics(2);
+
+            // With the fix: New exports should succeed because the hung request timed out
+            await().atMost(EXPORT_INTERVAL_MS * 2, TimeUnit.MILLISECONDS)
+                    .untilAsserted(() -> {
+                        int currentExports = SUCCESSFUL_EXPORTS.get();
+                        assertTrue(currentExports > baselineExports,
+                                String.format("Should have new successful exports after timeout. " +
+                                        "Baseline: %d, Current: %d", baselineExports, currentExports));
+                    });
+            LOG.infof("Test completed successfully - Total hanging requests: %d, Successful exports after recovery: %d",
```

Comment:
You need to refactor the whole test method for 2 reasons:

- it doesn't test the `quarkus.otel.exporter.otlp.metrics.timeout` configuration property, so I am worried to keep mentioning timeouts
- it assumes that metrics generated between `HANG_REQUESTS.set(true);` and ` HANG_REQUESTS.set(false);` are not delivered

but it works like this: `collected metrics -> (attempt) export -> retry export` so just because `(attempt) export` doesn't succeed, it doesn't mean that metrics are dropped

so you have:

- `generateMetrics(1);`
- `generateMetrics(2);`
- `generateMetrics(2);`

therefore you need to assert `Assertions.assertEquals(5, SUCCESSFUL_EXPORTS.get());` within some timeout (I used 2 minutes to test it but it only takes few seconds, try it).

Reply:
ok thanks @michalvavrik